### PR TITLE
wip: feat(menu): Add divider into `Menu` (like `DropdownDivider`)

### PR DESCRIPTION
**Pull request by rhengles**
### Work in Progress!!

I'm still missing the update on the changelog, and adding tests and docs.

I'm not very experienced with the details about changing Naive components, there are many details.

First I would like a heads up if this implementation is looking good. If it is, then I will proceed with the other necessary changes before this PR can be merged.

Thank you for this excellent library!

### CODE REVIEW

File: src/menu/src/Submenu.tsx [hunk 1]
```diff
@@ -16,18 +16,18 @@ import NMenuOptionContent from './MenuOptionContent'
 import { itemRenderer } from './utils'
 import { useMenuChild, useMenuChildProps } from './use-menu-child'
 import type { SubmenuInjection } from './use-menu-child'
-import { TreeNode } from 'treemate'
-import { MenuGroupOption, MenuOption, TmNode } from './interface'
+// import { TreeNode } from 'treemate'
```
Comment (07akioni) [inline src/menu/src/Submenu.tsx:19]:
```suggestion
```

File: src/menu/src/interface.ts [hunk 2]
```diff
@@ -33,7 +33,17 @@ export type MenuGroupOption =
...
   | (MenuGroupOptionBase & { label?: string | (() => VNodeChild) })
 
-export type TmNode = TreeNode<MenuOption, MenuGroupOption>
+export interface MenuIgnoredOption extends MenuOptionBase {
+  type: 'ignored' | 'divider'
+}
+
+export interface MenuDividerOption extends MenuOptionBase {
+  type: 'divider'
+}
```
Comment (07akioni) [inline src/menu/src/interface.ts:42]: Do not extends `MenuOptionsBase`. Other properties except `key` is useless.

File: src/menu/src/utils.ts [hunk 3]
```diff
@@ -4,15 +4,30 @@ import { keep, keysOf } from '../../_utils'
 import NMenuOptionGroup, { menuItemGroupProps } from './MenuOptionGroup'
 import NSubmenu, { submenuProps } from './Submenu'
 import NMenuOption, { menuItemProps } from './MenuOption'
-import { MenuOption, MenuGroupOption } from './interface'
+import NMenuOptionDivider from './MenuOptionDivider'
```
Comment (07akioni) [inline src/menu/src/utils.ts:7]:
```suggestion
import NMenuDivider from './MenuDivider'
```

**Comment by rhengles** [comment]
@07akioni I did some refactoring... please tell what you think.

File: src/menu/src/interface.ts [hunk 4]
```diff
@@ -9,13 +9,14 @@ export interface MenuOptionBase {
...
   [key: string]: unknown
   /** @deprecated */
   titleExtra?: string | (() => VNodeChild)
 }
 
 export interface MenuGroupOptionBase extends MenuOptionBase {
-  type: 'group'
+  type: 'group' | 'submenu'
   children: Array<MenuOption | MenuGroupOption>
 }
```
Comment (07akioni) [inline src/menu/src/interface.ts:21]: We'd better not mix 'group' and 'submenu' together in one type.

File: src/menu/src/MenuOption.tsx [hunk 5]
```diff
@@ -39,7 +39,10 @@ export default defineComponent({
     }
     function handleClick (e: MouseEvent): void {
       if (!mergedDisabledRef.value) {
-        NMenu.doSelect(props.internalKey, props.tmNode.rawNode)
+        NMenu.doSelect(
+          props.internalKey,
+          props.tmNode.rawNode as MenuOptionBase
```
Comment (07akioni) [inline src/menu/src/MenuOption.tsx:44]: It's `MenuOption`, not `MenuOptionBase`

File: src/menu/src/Menu.tsx [hunk 6]
```diff
@@ -198,11 +201,11 @@ export default defineComponent({
     })
 
     const treeMateRef = computed(() =>
-      createTreeMate<MenuOption, MenuGroupOption>(
+      createTreeMate<MenuOption, MenuGroupOption, MenuIgnoredOption>(
         props.items || props.options,
         {
           getKey (node) {
-            return node.key ?? node.name
+            return node.key ?? (node as MenuOptionBase).name
```
Comment (07akioni) [inline src/menu/src/Menu.tsx:208]: It's`MenuOption`, not `MenuOptionBase`.

File: src/dropdown/src/Dropdown.tsx [hunk 7]
```diff
@@ -55,7 +56,7 @@ DropdownIgnoredOption
     return node.key
   },
   getDisabled (node) {
-    return node.disabled === true
+    return (node as DropdownOptionBase).disabled === true
   },
```
Comment (07akioni) [inline src/dropdown/src/Dropdown.tsx:60]: It's `DropdownOption`, not `DropdownOptionBase`.

When doing force cast, you need to know, the type annotation should follow the runtime behavior.

A not accurate cast is worse than an `any` cast.

File: src/menu/src/interface.ts (hunk 4 above)
Comment (rhengles) [reply src/menu/src/interface.ts:21]: Well, yes, but in this case, there is a problem

I cannot extend DropdownGroupOption from MenuGroupOptionBase if it does not include this.

Should I make another type like the following?

```ts
// menu/src/interface.ts
export interface MenuGroupOptionBaseAbstract extends MenuOptionBase {
  type: 'group' | 'submenu'
  children: Array<MenuOption | MenuGroupOption>
}
export interface MenuGroupOptionBase extends MenuGroupOptionBaseAbstract {
  type: 'group'
}
// dropdown/src/interface.ts
export interface DropdownGroupOption extends MenuGroupOptionBaseAbstract {
  type: 'submenu'
}
```

**Comment by rhengles** [comment]
Sorry boss, I'm trying my best but I simply don't have enough experience with this codebase.

Please see doubt above about `'group' and 'submenu'`.

File: src/menu/src/interface.ts (hunk 4 above)
Comment (07akioni) [reply src/menu/src/interface.ts:21]: `type=submenu` is actually not required for `menu`, it's a compitable workaround.

Remove it seems to be okay.

Comment (rhengles) [reply src/menu/src/interface.ts:21]: Yes, it's ok for menu, but then dropdown cannot extend it. I just tried it and typescript show this error when building the package:

```
src/dropdown/src/interface.ts:12:18 - error TS2430: Interface 'DropdownGroupOption' incorrectly extends interface 'MenuGroupOptionBase'.
  Types of property 'type' are incompatible.
    Type '"submenu"' is not assignable to type '"group"'.

12 export interface DropdownGroupOption extends MenuGroupOptionBase {
                    ~~~~~~~~~~~~~~~~~~~
Found 1 error.
```

Comment (07akioni) [reply src/menu/src/interface.ts:21]: What if just remove 'submenu' from types?

Comment (rhengles) [reply src/menu/src/interface.ts]: @07akioni Alright, done! 👍

File: src/menu/src/interface.ts [hunk 8]
```diff
@@ -9,6 +9,7 @@ export interface MenuOptionBase {
   icon?: () => VNodeChild
   children?: Array<MenuOption | MenuGroupOption>
   extra?: string | (() => VNodeChild)
+  type?: 'group' | undefined
```
Comment (07akioni) [inline src/menu/src/interface.ts:12]:
```suggestion
  type?: 'group'
```

**Review by 07akioni — COMMENTED** [review]
Can you add a demo for divider?

**Comment by 07akioni** [comment]
The code seems to be okay.

**Comment by rhengles** [comment]
@07akioni Will do, thank you!

**Comment by 07akioni** [comment]
Try to solve conflict first?

**Comment by 07akioni** [comment]
@rhengles I've add it in https://github.com/TuSimple/naive-ui/pull/1636

**Pull request closed without merging**